File: DI_IOC_Observer/DI_IOC_Observer_PHP/AbstractPolygon.php
<?php
require_once "PolygonInterface.php";

abstract class AbstractPolygon implements SplSubject, PolygonInterface
{
    private $sideLength = 1;
    private $observers = [];
    protected $sides = 1;

    
    public function setSideLength($newSideLength) {
        if(!is_numeric($newSideLength) || $newSideLength <= 0) {
            throw new InvalidArgumentException('Side length must be number greater than 0.');
        }
        $this->sideLength = $newSideLength;
        $this->notify();
    }
    
    public function getSideLength() {
        return $this->sideLength;
    }
    
    public function getPerimeter() {
        return ($this->sides*$this->sideLength);
    }
    
    
    public function getNumberOfSides() {
        return $this->sides;
    }
    
    /////////////////////////////////////////////
    //SplSubject Interface bolierplate functions
    /////////////////////////////////////////////
    
    public function attach(SplObserver $observer) { 
        $oid = spl_object_hash($observer);
        if (!isset($this->observers[$oid])) {
            $this->observers[$oid] = $observer;
        }
        return $this;
    }
     
    public function detach(SplObserver $observer) {
        $oid = spl_object_hash($observer);
        if (isset($this->observers[$oid])) {
            unset($this->observers[$oid]);
        }
        return $this;
    }
     
    public function notify() {
        foreach ($this->observers as $observer) {    
            $observer->update($this);
        }
    }
}
?>

File: DI_IOC_Observer/DI_IOC_Observer_CS/DI_IOC_Observer_CS/AbstractPolygon.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DI_IOC_Observer_CS
{
    class AbstractPolygon : AbstractObservable<AbstractPolygon>, PolygonInterface
    {
        private double sideLength = 0.0f;
        protected int sides = 0;

        public AbstractPolygon() {
            sideLength = 1.0;
            sides = 1;
        }

        public double getPerimeter() {
            return ((double)sides * sideLength);
        }
        public void setSideLength(double newSideLength) {
            if(newSideLength <= 0.0) {
                throw new ArgumentOutOfRangeException("newSideLength" , "Side length must be a number greater than 0.");
            }
            sideLength = newSideLength;
            this.Notify(this);
        }
        public double getSideLength() {
            return sideLength;
        }
        public int getNumberOfSides() {
            return sides;
        }
    }
}